### ISSUE TITLE
Feature Request: optional caching of artifacts downloaded

### ISSUE THREAD

**Issue by ketzacoatl** [hashicorp/nomad#5217]
When downloading artifacts to run a task, there is the startup cost to download the artifact. When a task is updated / restarted in some ways, when we end up with a new task allocation, the artifacts must be re-downloaded. This is especially difficult when using the `exec`/`raw_exec` drivers, but also applies to the other drivers as well. This is as of `v0.8.6`.

If downloaded files were kept in a local file cache, of some limited size or age, nomad could check to see if a cached file matches the checksum from the task and prefer to use it before re-downloading the artifact (if the user says that's ok to do).

**Comment by cpumanaz** [comment]
Nomad should provide for using the host docker registry.  I have a workaround which is to create localhost:5000 registry and use that. It does speed "download" times but still duplicates image data.

Current Behavior:  
Nomad downloads all images from public/private registry. (even if images are pulled locally through docker) Nomad caches only images it allocates.

Desired Behavior:
Nomad follows standard Docker behavior. Use the local Docker cache instead of an internal one. If always pull is chosen in the docker driver then Nomad will check the remote registry and pull a new copy into the Docker cache which it will allocate. If purging or cleanup is specified, Nomad can purge/clean from the local docker cache. (docker image rm)

**Comment by ketzacoatl** [comment]
I think that is something to consider as a separate feature. I do not think nomad should run a docker registry, and that is not in scope for my request. My request aims to focus on the `artifact` stanza. How the localhost handles docker images is a different subsystem, so I think it would be best to consider that in a separate ticket.

**Comment by shantanugadgil** [comment]
Just a thought, this might need agent specific configuration to set the artifact directory?

**Comment by zhiguangwang** [comment]
This feature would be really useful.

I'm using Nomad `exec` driver to run dedicated servers of Unreal Engine 4, and each server artifact (.tar.gz) is 400+MB, so to deploy 20 server instances to one Nomad client machine would mean downloading and extracting 8GB+ of data.

Further more, it would be even better if Nomad could reuse the extracted artifact files (read-only) across all allocations of the same job on the same machine, which would minimize disk consumption  and maximize instance startup speed.

**Comment by prologic** [comment]
We wrote our own "artifact server" that is run as a Nomad Job using the `exec` driver. Without Nomad internally caching artifact(s) we have a chicken/egg problem whereby we can't upgrade our "artifact server" unless we spin up the new one in parallel pointing at URIs that match the old server before shutting the old one down.

**Comment by alexboboc** [comment]
This would be very useful. We're running hundreds of tasks that require downloading the latest version of a binary through the `raw_exec` driver and having a separate copy for each job is just not optimal. 
Being able to cache downloaded artifacts to avoid redownloading identical copies would be a much better option.

**Comment by 02strich** [comment]
@alexboboc do you have any concerns with the disk size consumed by the multiple copies or solely the transfer of the data?

**Comment by spaiz** [comment]
Also asking for this feature. We are using nomad for local development, and stop/run tasks often. There is no need to wait each time to download an already existing image.

**Comment by freef4ll** [comment]
Would like to also say that cache of artifact stanza when using `exec`/`raw_exec` is very useful for production and local development.

There are workarounds like using http and s3 caching via [nginx-s3-gateway](https://github.com/nginxinc/nginx-s3-gateway), but this adds to complexity.

**Comment by stenh0use** [comment]
This feature seems worthwhile to me. I think setting `ephemeral_disk.sticky` to `true` on a job and saving the artifact to `alloc/data/` would provide some job level caching, unless nomad forces the go-getter library to download the artifact again. My testing with raw go-getter cli commands confirms that the library does not re-download the artifact.
